BUG: Fixes pd.merge issue with columns of dtype numpy.uintc on windows

AbhishekChaudharii opened the pull request:
This PR fixes the issue encountered when using pd.merge with columns of dtype numpy.uintc on windows.
Added a relevant pytest case to validate correct behavior.

- [x] closes #58713 
- [x] [Tests added and passed
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

AbhishekChaudharii committed:
bug fix for numpy.uintc in merge operations on windows

Added pytest test case to verify correct behavior with numpy.uintc dtype

AbhishekChaudharii committed:
Formatting changes after running pre-commit

mroeschke commented on pandas/tests/reshape/merge/test_merge.py:
```diff
@@ -1843,6 +1843,20 @@ def test_merge_empty(self, left_empty, how, exp):
         tm.assert_frame_equal(result, expected)
 
+    def test_merge_with_uintc_columns(dataframes_with_uintc):
+        """To test if pd.merge works with numpy.uintc on windows"""
+
```
```suggestion
```
We don't use docstrings in our tests

mroeschke commented on pandas/tests/reshape/merge/test_merge.py:
```diff
@@ -1843,6 +1843,20 @@ def test_merge_empty(self, left_empty, how, exp):
+    def test_merge_with_uintc_columns(dataframes_with_uintc):
+        """To test if pd.merge works with numpy.uintc on windows"""
+
+        df1 = pd.DataFrame({"a": ["foo", "bar"], "b": np.array([1, 2], dtype=np.uintc)})
```
Do we have tests with `np.intc` as well?

mroeschke reviewed (changes requested):
Could you also add a whatsnew note in `v3.0.0.rst` in the `Reshaping` section?

AbhishekChaudharii committed:
Merge branch 'pandas-dev:main' into issue_58713

AbhishekChaudharii committed:
Added tests for numpy.intc

AbhishekChaudharii committed:
Merge branch 'issue_58713' of https://github.com/AbhishekChaudharii/pandas into issue_58713

AbhishekChaudharii committed:
added whatsnew note

AbhishekChaudharii committed:
pre-commit automatic changes and also made changes to test_merge.py file to make pandas namespace consistent

AbhishekChaudharii commented:
Hi @mroeschke , I have added the whatsnew note and included current issue #58713 and also issue #60091 as this PR fixes both of them. I have also added 2 more test cases which includes `np.intc` and reproducible example from issue #60091

AbhishekChaudharii committed:
removed comment

mroeschke commented on doc/source/whatsnew/v3.0.0.rst:
```diff
@@ -592,7 +592,6 @@ Performance improvements
 - Performance improvement in :meth:`RangeIndex.take` returning a :class:`RangeIndex` instead of a :class:`Index` when possible. (:issue:`57445`, :issue:`57752`)
 - Performance improvement in :func:`merge` if hash-join can be used (:issue:`57970`)
 - Performance improvement in :meth:`CategoricalDtype.update_dtype` when ``dtype`` is a :class:`CategoricalDtype` with non ``None`` categories and ordered (:issue:`59647`)
-- Performance improvement in :meth:`DataFrame.astype` when converting to extension floating dtypes, e.g. "Float64" (:issue:`60066`)
```
Don't think this should have been removed

AbhishekChaudharii replied:
Sorry It accidentally got deleted while running a git command. Thank you for catching it

mroeschke commented on doc/source/whatsnew/v3.0.0.rst:
```diff
@@ -737,6 +736,8 @@ Reshaping
 - Bug in :meth:`DataFrame.join` when a :class:`DataFrame` with a :class:`MultiIndex` would raise an ``AssertionError`` when :attr:`MultiIndex.names` contained ``None``. (:issue:`58721`)
 - Bug in :meth:`DataFrame.merge` where merging on a column containing only ``NaN`` values resulted in an out-of-bounds array access (:issue:`59421`)
 - Bug in :meth:`DataFrame.unstack` producing incorrect results when ``sort=False`` (:issue:`54987`, :issue:`55516`)
+- Bug in :meth:`DataFrame.merge` when merging two dataframes with column dtype as numpy.intc resulting in ValueError: Buffer dtype mismatch, Only on windows (:issue:`60091`)
+- Bug in :meth:`DataFrame.merge` when merging two dataframes with column dtype as numpy.uintc resulting in KeyError: <class 'numpy.uintc'> ,Only on windows (:issue:`58713`)
```
```suggestion
- Bug in :meth:`DataFrame.merge` when merging two :class:`DataFrame` on ``intc`` or ``uintc`` types on Windows (:issue:`60091`, :issue:`58713`)
```

AbhishekChaudharii committed:
Merge branch 'pandas-dev:main' into issue_58713

AbhishekChaudharii committed:
added the deleted whatsnew note back

AbhishekChaudharii committed:
better whatsnew note

Co-authored-by: Matthew Roeschke <10647082+mroeschke@users.noreply.github.com>

mroeschke commented:
Thanks @AbhishekChaudharii
